lib/private.h

```diff
@@ -129,7 +129,7 @@ struct vfu_ctx {
 
     uint32_t                irq_count[VFU_DEV_NUM_IRQS];
     vfu_irqs_t              *irqs;
-    bool                    ready;
+    bool                    realized;
```

Comment:
IMO we don't need to store this piece of information, if we decide to do so then why not call `vfu_ctx_realize` automatically if it hasn't been done so.

Reply:
The only use of this bool is to return EINVAL from vfu_drive_ctx() and vfu_ctx_poll() if the device is not realized(). I think instead of calling realized() everytime its better to enforce user and fail if device not realized(). 

Or we can get rid of this bool and process_context() may fail weirdly is device is not realized() by user.

Reply:
drive() and poll() will also fail weirldy if the user hasn't set up regions or IRQs, I don't think we can protect against misuse. We should document very well the API and provide samples to avoid such problems.

Reply:
Yes but I see that realized() is the step which is after the device setup i.e regions(), IRQ etc. So realized() can always return errors and user can pre-do setup accordingly.
But this is debatable and we can just document and let process_context() behave with un-realized() device.

Reply:
Realize will not necessarily return errors if the user didn't e.g. configure a region they were plan to use, right?

Reply:
Yes we cannot predict what user wants :) and my point was not even in that direction. We can only do some sanity checks which requires the setup step to be completed. 
Ok lets take discussion back to the options,
1. Remove the bool and assume user realized() the device and just continue with drive() or poll().
2. Keep bool and always realized() in drive() and poll()
Let me know which works for you.

Reply:
I'm fine with either options, @jlevon what do you think?

Reply:
I think what Swapnil has here already is fine personally.

Reply:
OK